### PR TITLE
Extend EXP1 memory range and allow mounting rom

### PR DESCRIPTION
**Pull request by johnbaumann**
Increased EXP1 mapped memory range from 64kb to 512kb. Added option to mount rom from file. Added settings for rom file, rom path, as well as bios path to ease navigation.

Add basic PIO cart emulation and config menu for mounting cart sizes up to 256kb.

### CODE REVIEW

File: src/core/psxmem.cc [hunk 1]
```diff
@@ -92,35 +92,43 @@ int PCSX::Memory::init() {
     m_writeLUT = (uint8_t **)calloc(0x10000, sizeof(void *));
 
     m_psxM = (uint8_t *)calloc(0x00800000, 1);
-    m_psxP = (uint8_t *)calloc(0x00010000, 1);
+    m_psxP = (uint8_t *)calloc(0x00080000, 1);
```
Comment (wheremyfoodat) [inline src/core/psxmem.cc:95]: It's probably best to start doing away with the old names for the memory ranges. Nobody is going to read `m_psxP` and think "Oh yes this is PIO memory". Something like `m_pio` is just as short and much more clear.

File: src/core/psxmem.cc [hunk 2]
```diff
@@ -92,35 +92,43 @@ int PCSX::Memory::init() {
...
 
-    if (m_readLUT == NULL || m_writeLUT == NULL || m_psxM == NULL || m_psxP == NULL || m_psxH == NULL) {
+    if (m_readLUT == NULL || m_writeLUT == NULL || m_psxM == NULL || m_psxP == NULL || m_psxR == NULL || m_psxH == NULL) {
         g_system->message("%s", _("Error allocating memory!"));
         return -1;
     }
 
-    // MemR
-    m_readLUT[0x1f00] = (uint8_t *)m_psxP;
+    // MemR / MemW
```
Comment (wheremyfoodat) [inline src/core/psxmem.cc:104]: It's also probably best to do away with the old code's tendency of having comments for self-explanatory code written in half-legible English. Not much necessity in specifying "MemR/MemW" when it's obviously made R/W below

File: src/core/psxmem.h [hunk 3]
```diff
@@ -45,10 +45,10 @@ namespace PCSX {
...
 class Memory {
   public:
-    uint8_t *m_psxM = NULL;  // Kernel & User Memory (8 Meg)
-    uint8_t *m_psxP = NULL;  // Parallel Port (64K)
-    uint8_t *m_psxR = NULL;  // BIOS ROM (512K)
-    uint8_t *m_psxH = NULL;  // Scratch Pad (1K) & Hardware Registers (8K)
+    uint8_t *m_main = NULL;        // Kernel & User Memory (8 Meg)
+    uint8_t *m_exp1 = NULL;        // Expansion Region 1 (ROM/RAM) / Parallel Port (512K)
+    uint8_t *m_bios = NULL;        // BIOS ROM (512K)
+    uint8_t *m_scratchpad = NULL;  // Scratch Pad (1K) & Hardware Registers (8K)
```
Comment (nicolasnoble) [inline src/core/psxmem.h:51]: Renaming this to "scratchpad" hides the fact this also contains a copy of hardware registers. The H used to stand for Hardware. Could use `m_hard` to keep alignment?

File: src/core/psxmem.cc [hunk 4]
```diff
@@ -133,65 +140,86 @@ void PCSX::Memory::reset() {
...
+            PCSX::g_system->printf(_("Loaded BIOS: %s\n"), biosPath.string());
         }
     }
 
-    if (!f->failed()) {
-        f->read(m_psxR, bios_size);
-        f->close();
-        PCSX::g_system->printf(_("Loaded BIOS: %s\n"), biosPath.string());
+    // Load EXP1
+    {
```
Comment (nicolasnoble) [inline src/core/psxmem.cc:190]: Maybe let's keep a boolean somewhere to indicate the exp1 is loaded with something. If not loaded, we should nuke the corresponding LUTs.

File: src/core/sstate.h [hunk 5]
```diff
@@ -44,7 +44,7 @@ typedef Protobuf::MessageField<Thumbnail, TYPESTRING("thumbnail"), 2> ThumbnailF
 
 typedef Protobuf::FieldPtr<Protobuf::FixedBytes<0x00800000>, TYPESTRING("ram"), 1> RAM;
 typedef Protobuf::FieldPtr<Protobuf::FixedBytes<0x00080000>, TYPESTRING("rom"), 2> ROM;
-typedef Protobuf::FieldPtr<Protobuf::FixedBytes<0x00010000>, TYPESTRING("parallel"), 3> Parallel;
+typedef Protobuf::FieldPtr<Protobuf::FixedBytes<0x00080000>, TYPESTRING("parallel"), 3> Parallel;
```
Comment (nicolasnoble) [inline src/core/sstate.h:47]: It's fine to rename this one to Exp1-style names.

File: src/core/sstate.h [hunk 6]
```diff
@@ -44,7 +44,7 @@ typedef Protobuf::MessageField<Thumbnail, TYPESTRING("thumbnail"), 2> ThumbnailF
...
 typedef Protobuf::FieldPtr<Protobuf::FixedBytes<0x00010000>, TYPESTRING("hardware"), 4> HardwareMemory;
 typedef Protobuf::Message<TYPESTRING("Memory"), RAM, ROM, Parallel, HardwareMemory> Memory;
```
Comment (nicolasnoble) [inline src/core/sstate.h:49]: Probably want to also save the boolean about wether or not the exp1 file is loaded.

File: src/gui/gui.cc [hunk 7]
```diff
@@ -1266,32 +1266,32 @@ in Configuration->Emulation, restart PCSX-Redux, then try again.)"));
...
+            m_hwrEditor.draw(g_emulator->m_mem->m_scratchpad + 4 * 1024, 8 * 1024, 0x1f801000);
         }
         counter++;
         if (m_biosEditor.show) {
             ImGui::SetNextWindowPos(ImVec2(520, 30 + 10 * counter), ImGuiCond_FirstUseEver);
             ImGui::SetNextWindowSize(ImVec2(484, 480), ImGuiCond_FirstUseEver);
-            m_biosEditor.draw(g_emulator->m_mem->m_psxR, 512 * 1024, 0xbfc00000);
+            m_biosEditor.draw(g_emulator->m_mem->m_bios, 512 * 1024, 0xbfc00000);
         }
         counter++;
```
Comment (nicolasnoble) [inline src/gui/gui.cc:1296]: Might want to add an editor for the exp1 file.

File: src/gui/gui.cc [hunk 8]
```diff
@@ -1710,6 +1711,18 @@ faster by not displaying the logo.)"));
...
+        ImGui::InputText(_("EXP1 file"), const_cast<char*>(reinterpret_cast<const char*>(exp1.c_str())), exp1.length(),
+                         ImGuiInputTextFlags_ReadOnly);
+        ImGui::SameLine();
+        selectEXP1Dialog = ImGui::Button("...##EXP1");
+        ImGui::SameLine();
+        if (ImGui::Button(_("Clear"))) {
+            settings.get<Emulator::SettingEXP1>().value = "";
+            changed = true;
+        }
+
```
Comment (nicolasnoble) [inline src/gui/gui.cc:1725]: Might want to add a button to "unload" or "unplug" the exp1 cart, and toggle the "is loaded" boolean accordingly, and rebuild the LUTs.

File: src/core/psxmem.h (hunk 3 above)
Comment (wheremyfoodat) [reply src/core/psxmem.h:51]: m_wram over `m_main` probably makes more sense too. Also miiiight want to squash some of these commits.

**Comment by asmblur** [comment]
Did I not add many of these features 3y ago? 🥴

File: src/core/psxmem.cc [hunk 9]
```diff
@@ -274,6 +325,43 @@ uint32_t PCSX::Memory::read32(uint32_t address) {
...
+    if (!L.isnil()) {
+        try {
+            // Call the function. Technically we should empty
+            // the stack afterward to clean potential return
+            // values from the Lua code, this is a bug in the gui code.
+            const int top = L.gettop();
+            L.push(lua_Number(address));
+            L.push(lua_Number(size));
+            nresult = L.pcall(2);
+        } catch (...) {
```
Comment (nicolasnoble) [inline src/core/psxmem.cc:344]: We may want to discard any results here that's above 1.

Something like
```c++
for (int n = 1; n < nresult; n++) {
    L.pop();
}
```

File: src/core/psxmem.cc [hunk 10]
```diff
@@ -274,6 +325,43 @@ uint32_t PCSX::Memory::read32(uint32_t address) {
...
+        }
+    } else {
+        // This pops the nil from the stack.
+        L.pop();
+    }
+
+    return nresult;
+}
+
+void PCSX::Memory::sendWriteToLua() {}
```
Comment (nicolasnoble) [inline src/core/psxmem.cc:363]: This looks defined but not used ?

File: src/core/psxmem.h [hunk 11]
```diff
@@ -136,9 +136,16 @@ class Memory {
...
+    void LoadEXP1FromFile(std::filesystem::path rom_path);
+    void sendWriteToLua();
+    int sendReadToLua(uint32_t address, size_t size);
+
   private:
+    
     int m_writeok = 1;
     uint32_t m_biosCRC = 0;
+
+    std::vector<std::string> m_memoryErrors;
```
Comment (nicolasnoble) [inline src/core/psxmem.h:148]: This doesn't look used ?

File: src/gui/widgets/assembly.cc [hunk 12]
```diff
@@ -538,25 +538,25 @@ settings, otherwise debugging features may not work.)");
...
-                    nextCode = *reinterpret_cast<uint32_t*>(m_memory->m_psxM + addr + 4);
+                    nextCode = *reinterpret_cast<uint32_t*>(m_memory->m_wram + addr + 4);
                 }
                 base = m_ramBase;
             } else if (addr < 0x00810000) {
                 section = "PAR";
                 addr -= 0x00800000;
-                code = *reinterpret_cast<uint32_t*>(m_memory->m_psxP + addr);
+                code = *reinterpret_cast<uint32_t*>(m_memory->m_exp1 + addr);
                 if (addr <= 0x0000fff8) {
```
Comment (nicolasnoble) [inline src/gui/widgets/assembly.cc:550]: This becomes technically wrong, due to the size change of the exp1 section. But then again, I need to heavily rewrite this code.

File: src/core/pio-cart.h [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+        }
+    }
+
+    void setLuts();
+
+    bool getSwitch() { return m_switchOn; }
+    void setSwitch(bool on) { m_switchOn = on & 1; }
+
+    uint8_t read8(uint32_t address);
+    uint16_t read16(uint32_t address) { return 0xffff; }      // 2 reads?
```
Comment (nicolasnoble) [inline src/core/pio-cart.h:49]: Might as well do these here? Like
```c++
uint16_t read16(uint32_t address) {
    uint16_t b1 = read8(address);
    uint16_t b2 = read8(address + 1);
    return b1 | (b2 << 16);
}
```
(etc)

File: src/core/pio-cart.h [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+                m_busCycle = 0;
+            }
+
+            void resetFlash() {
+                resetCommandBuffer();
+                m_dataProtectEnabled = true;
+                m_pageWriteEnabled = false;
+            }
+
+            void sectorErase() {}
```
Comment (nicolasnoble) [inline src/core/pio-cart.h:122]: We might need the address of the sector? And it's going to memset it to 0xff.

File: src/core/psxmem.cc (hunk 10 above)
Comment (johnbaumann) [reply src/core/psxmem.cc:363]: Woops. I had removed this but it came back after I had that commit/push bring back everything and had to re-force push.

File: src/core/psxmem.h (hunk 11 above)
Comment (johnbaumann) [reply src/core/psxmem.h:148]: This too

File: src/core/pio-cart.h (hunk 14 above)
Comment (johnbaumann) [reply src/core/pio-cart.h:122]: Looks like this was from one of the earlier flash chips I investigated. I did go ahead and implemented the chip-erase command

**Review by nicolasnoble — APPROVED** [review]

**Review by nicolasnoble — APPROVED** [review]

**Pull request merged**